File: NirvanaSMS.Api.Java/src/main/java/org/nirvanasms/api/core/utility/PhoneNumberFormatter.java
package org.nirvanasms.api.core.utility;

import org.nirvanasms.api.core.base.FormatterBase;
import org.nirvanasms.api.core.entity.PhoneNumber;

public class PhoneNumberFormatter extends FormatterBase<PhoneNumber, PhoneNumber> {
	ValidNumber providerValidNumberType = ValidNumber.Type4;

	@Override
	public PhoneNumber format(PhoneNumber phoneNumber) {
		phoneNumber.setNumber(phoneNumber.getNumber().trim());
		ValidNumber numberType = IsPhoneNumber(phoneNumber.getNumber());
		boolean isValid = numberType != ValidNumber.NonValid;
		boolean isProviderPhoneNumber = isProviderPhoneNumber(numberType);
		if (isValid && isProviderPhoneNumber)
			phoneNumber.setValid(true);
		else if (isValid && !isProviderPhoneNumber) {
			// changing number to valid format
			switch (numberType) {
			case Type1:
				phoneNumber.setNumber(changeToValidFormat(phoneNumber.getNumber()));
				break;
			case Type2:
				phoneNumber.setNumber(changeToValidFormat(phoneNumber.getNumber()));
				break;
			case Type3:
				phoneNumber.setNumber(changeToValidFormat(phoneNumber.getNumber()));
				break;
			default:
				break;
			}
			phoneNumber.setValid(true);
		} else
			phoneNumber.setValid(false);
		return phoneNumber;
	}

	private ValidNumber IsPhoneNumber(String input) {
		String[] patterns = { "([\\+]90?[ ])(\\(5+[0-9]{2}\\)[ ])([0-9]{3})([\\-])([0-9]{2})([\\-])([0-9]{2})", // Type
																												// 1
				"([\\+]90?)([ ]?)(\\([0-9]{3}\\))([ ]?)([0-9]{3})(\\s*[\\-]?)([0-9]{2})(\\s*[\\-]?)([0-9]{2})", // Type
																												// 2
				"(([\\+]90?)|([0]?))([ ]?)(\\([0-9]{3}\\))([ ]?)([0-9]{3})(\\s*[\\-]?)([0-9]{2})(\\s*[\\-]?)([0-9]{2})", // Type
																															// 3
				"(([\\+]90?)|([0]?))([ ]?)((\\([0-9]{3}\\))|([0-9]{3}))([ ]?)([0-9]{3})(\\s*[\\-]?)([0-9]{2})(\\s*[\\-]?)([0-9]{2})" // Type
																																		// 4
																																		// //
																																		// -
																																		// //
																																		// Provider
		};

		for (int i = 0; i < patterns.length; i++) {
			String item = patterns[i];
			if (input.matches(item))
				return ValidNumber.values()[i];
		}

		return ValidNumber.NonValid;
	}

	private boolean isProviderPhoneNumber(ValidNumber number) {
		return number == providerValidNumberType;
	}

	private String changeToValidFormat(String type3Number) {
		String type4Result = "";
		type4Result = type3Number.replace("(", "").replace(")", "");
		return type4Result;
	}

	enum ValidNumber {
		NonValid, Type1, Type2, Type3, Type4
	}

}

File: NirvanaSMS.Api.CSharp/Core/Utility/PhoneNumberFormatter.cs
using NirvanaSMS.Api.Core.Base;
using NirvanaSMS.Api.Core.Entity;
using System.Text.RegularExpressions;

namespace NirvanaSMS.Api.Core.Utility
{
    public class PhoneNumberFormatter : FormatterBase<PhoneNumber, PhoneNumber>
    {
        ValidNumber providerValidNumberType = ValidNumber.Type4;

        public override PhoneNumber Format(PhoneNumber phoneNumber)
        {
            phoneNumber.Number.Trim();
            ValidNumber numberType = IsPhoneNumber(phoneNumber.Number);
            bool isValid = numberType != ValidNumber.NonValid;
            bool isProviderPhoneNumber = IsProviderPhoneNumber(numberType);
            if (isValid && isProviderPhoneNumber)
                phoneNumber.IsValid = true;
            else if (isValid && !isProviderPhoneNumber)
            {
                //changing number to valid format
                switch (numberType)
                {
                    case ValidNumber.Type1:
                        phoneNumber.Number = changeToValidFormat(phoneNumber.Number);
                        break;
                    case ValidNumber.Type2:
                        phoneNumber.Number = changeToValidFormat(phoneNumber.Number);
                        break;
                    case ValidNumber.Type3:
                        phoneNumber.Number = changeToValidFormat(phoneNumber.Number);
                        break;
                }
                phoneNumber.IsValid = true;
            }
            else
                phoneNumber.IsValid = false;
            return phoneNumber;
        }

        private ValidNumber IsPhoneNumber(string input)
        {
            string[] patterns = new string[] {
                @" ([\+]90?[ ])(\(5+[0-9]{2}\)[ ])([0-9]{3})([\-])([0-9]{2})([\-])([0-9]{2})", //Type 1
                @"([\+]90?)([ ]?)(\([0-9]{3}\))([ ]?)([0-9]{3})(\s*[\-]?)([0-9]{2})(\s*[\-]?)([0-9]{2})", //Type 2
                @"(([\+]90?)|([0]?))([ ]?)(\([0-9]{3}\))([ ]?)([0-9]{3})(\s*[\-]?)([0-9]{2})(\s*[\-]?)([0-9]{2})", //Type 3
                @"(([\+]90?)|([0]?))([ ]?)((\([0-9]{3}\))|([0-9]{3}))([ ]?)([0-9]{3})(\s*[\-]?)([0-9]{2})(\s*[\-]?)([0-9]{2})" //Type 4 - Provider
            };

          for(int i = 0; i < patterns.Length; i++)
            {
                string item = patterns[i];
                if (Regex.Match(input, item, RegexOptions.IgnoreCase).Success)
                    return (ValidNumber)i;
            }

            return ValidNumber.NonValid;
        }
        //905 076982056
        private bool IsProviderPhoneNumber(ValidNumber number)
        {
            return number == providerValidNumberType;
        }

        string changeToValidFormat(string type3Number)
        {
            string type4Result = "";
            type4Result = type3Number.Replace("(", "").Replace(")", "");
            return type4Result;
        }

        enum ValidNumber
        {
            NonValid = -1,
            Type1 = 0,
            Type2 = 1,
            Type3 = 2,
            Type4 = 3,
        }
    }
}
